### PR TITLE
PERF: avoid unnecessary reordering in MultiIndex._reorder_indexer

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v1.6.0.rst` file if fixing a bug or adding a new feature.

`MultiIndex._reorder_indexer` contains logic to determine if indexer reordering is required. This PR expands the logic a bit to identify additional cases where reordering is not necessary.

ASVs: 

```
       before           after         ratio
     [4b645ae5]       [07679a60]
     <main>           <multiindex-need-sort>
-        5.07±1ms          726±9μs     0.14  indexing.MultiIndexing.time_loc_all_bool_indexers(True)
-        8.13±2ms         709±20μs     0.09  indexing.MultiIndexing.time_loc_all_bool_indexers(False)
```

### CODE REVIEW

**Review by datapythonista — APPROVED** [review]
Thanks @lukemanley, looks good to me.

Can you merge master into your branch? I think CI problems should be fixed now.

File: pandas/core/indexes/multi.py [hunk 1]
```diff
@@ -3466,22 +3466,35 @@ def _reorder_indexer(
...
-                        k_codes = self.levels[i].get_indexer(k)
-                        k_codes = k_codes[k_codes >= 0]  # Filter absent keys
-                        # True if the given codes are not ordered
-                        need_sort = (k_codes[:-1] > k_codes[1:]).any()
-                elif isinstance(k, slice) and k.step is not None and k.step < 0:
+
+        # check if sorting is necessary
+        need_sort = False
+        for i, k in enumerate(seq):
+            if com.is_null_slice(k) or com.is_bool_indexer(k) or is_scalar(k):
```
Comment (mroeschke) [inline pandas/core/indexes/multi.py:3473]: Do we have unit tests for all these new branches that short circuit?

Comment (lukemanley) [reply pandas/core/indexes/multi.py:3473]: I'm pretty there are tests that hit these in one way or another. Regardless, I just added test_get_locs_reordering that explicitly tests the ordering effects of: null slices, bool indexers, scalars, and lists of length 1.

**Review by mroeschke — APPROVED** [review]

File: pandas/core/indexes/multi.py (hunk 1 above)
Comment (mroeschke) [reply pandas/core/indexes/multi.py:3473]: Great thanks!

**Comment by mroeschke** [comment]
Thanks @lukemanley

**Pull request merged**